### DISCUSSION TITLE
Loading images as contiguous memory

### DISCUSSION THREAD

**Discussion by ThomasMiz** [SixLabors/ImageSharp#1258] [Q&A] [answered]
I've been using ImageSharp for quite some time to load images in a [graphics library I've been developing](https://github.com/ThomasMiz/TrippyGL).

ImageSharp has been a great cross-platform solution for decoding images in different formats and getting a contiguous span of data I can pass to OpenGL and load into a texture.

However, in a recent update, the image.GetPixelSpan() function was removed.

First, I tried using image.GetPixelRowSpan(0) and that seemed to work, and [looking at this commit from ImageSharp.Textures](https://github.com/SixLabors/ImageSharp.Textures/commit/bbec8b7f8483643902cbac380123b7d5b360be0b) this seemed to be the way to go.

But then I saw the image.TryGetSinglePixelSpan() function and sure enough, in my test project that loads a single texture I can get a Span<Rgba32> with the entire image as desired. Unfortunately for me, the syntax of this function clearly implies that the memory of images can no longer be guaranteed to be contiguous, which is something I need.

Is there a good solution to this? It would sure be nice to have a way to ensure an Image's memory is contiguous and I imagine I'm not the only one having this issue.

**Comment by antonfirsov** [comment, marked as answer]
Discontiguous buffers have been introduced in #1109 to enable handling **very** large images. The size of a single contiguous buffer is limited by `ArrayPoolMemoryAllocator.BufferCapacityInBytes`. It's current (arbitrary) default value is `int.MaxValue/4` which means that an `Image<Rgba32>` below the size of ~134 megapixels (`~ 12k x 11k` pixels) will always fit into a single buffer with the default setup. However, we may decide to lower this value, so if you want to make sure your images fit into a single buffer, I'd suggest to create a custom `ArrayPoolMemoryAllocator` instance, and assign it to `Configuration.Default.MemoryAllocator` (or to the same property on your custom `Configuration` instance).

Unfortunately, at the moment there's no other way than to use the full constructor of `ArrayPoolMemoryAllocator`, feeding the defaults to preceding parameters:
https://github.com/SixLabors/ImageSharp/blob/ab8ab3992a93f6682f0c48a591e5df84151081fe/src/ImageSharp/Memory/Allocators/ArrayPoolMemoryAllocator.cs#L85-L90

**Reply by ThomasMiz** [reply]
Does this mean, assuming `ArrayPoolMemoryAllocator.BufferCapacityInBytes` is `int.MaxValue/4`, that I can guarantee that any `Image<Rgba32>` below ~134 megapixels will be contiguous in memory?

There's no such thing as a non-contiguous 1024x1024 `Image<Rgba32>`?

**Reply by antonfirsov** [reply]
Exactly.

We may lower this threshold if we find out that more intense usage of noncontigous blocks improves throughput for the most common resize scenarios. But I think we'll never go below 1MP.

**Reply by ThomasMiz** [reply]
Alright, thanks!

By the way, the SixLabors set of libraries are incredible, so special thanks to everyone who worked on these for their amazing work! Open source projects like these hold our personal projects together!
